src/python-common/ceph/rgw/rgwam_core.py

```diff
@@ -386,101 +430,125 @@ def zone_op(self):
-
+    def create_normal_user(self, zg, zone, uid=None):
         try:
             user_info = self.user_op().create(zone, zg, uid=uid, is_system=False)
+            user = RGWUser(user_info)
+            logging.info('Created regular user: %s' % user.uid)
+            return user
         except RGWAMException as e:
             raise RGWAMException('failed to create user', e)
 
-        user = RGWUser(user_info)
-
-        logging.info('Created regular user: %s' % user.uid)
-
-        eps = endpoints.split(',')
-        ep = ''
-        if len(eps) > 0:
-            ep = eps[0]
-            if start_radosgw:
-                o = urlparse(ep)
-                svc_id = realm_name + '.' + zone_name
-                self.env.mgr.apply_rgw(svc_id, realm_name, zone_name, o.port)
-
-        realm_token = RealmToken(realm_id, ep, sys_user.uid, sys_access_key, sys_secret)
-
-        logging.info(realm_token.to_json())
+    def update_period(self, realm, zg, zone=None):
+        try:
+            period_info = self.period_op().update(realm, zg, zone, commit=True)
+            period = RGWPeriod(period_info)
+            logging.info('Period: ' + period.id)
+        except RGWAMCmdRunException as e:
+            raise RGWAMException('failed to update period', e)
 
-        realm_token_b = realm_token.to_json().encode('utf-8')
-        return (0, 'Realm Token: %s' % base64.b64encode(realm_token_b).decode('utf-8'), '')
+    def realm_bootstrap(self, rgw_spec, start_radosgw=True):
+
+        realm_name = rgw_spec.rgw_realm
+        zonegroup_name = rgw_spec.rgw_zonegroup
+        zone_name = rgw_spec.rgw_zone
+
+        # Some sanity checks
+        if realm_name in self.realm_op().list():
+            raise RGWAMException(f'Realm {realm_name} already exists')
+        if zonegroup_name in self.zonegroup_op().list():
+            raise RGWAMException(f'ZonegroupOp {zonegroup_name} already exists')
+        if zone_name in self.zone_op().list():
+            raise RGWAMException(f'Zone {zone_name} already exists')
+
+        # Create RGW multisite entities and update the period
+        realm = self.create_realm(realm_name)
+        zonegroup = self.create_zonegroup(realm, zonegroup_name, zonegroup_is_master=True)
+        zone = self.create_zone(realm, zonegroup, zone_name, zone_is_master=True)
+        self.update_period(realm, zonegroup)
+
+        # Create system user, normal user and update the master zone
+        sys_user = self.create_system_user(realm, zonegroup, zone)
+        access_key = sys_user.keys[0].access_key
+        secret = sys_user.keys[0].secret_key
+        self.zone_op().modify(zone, zonegroup, None, None, access_key, secret)
+        self.update_period(realm, zonegroup)
+        self.create_normal_user(zonegroup, zone)
+
+        if start_radosgw:
+            # Instruct the orchestrator to start RGW daemons, asynchronically, this will
+            # call back the rgw module to update the master zone with the corresponding endpoints
```

Comment:
going of the code in cephadm, we only update these endpoints if `rgw_spec.update_endpoints and rgw_spec.rgw_realm_token is not None`. Do we have those guarantees with the spec we have here (we might, legitimately asking)?